tests/test_noise.py

```diff
@@ -130,3 +138,201 @@ def pinkish_spectrum(x, y):
+)
+@pytest.mark.parametrize("dtype", DTYPES, ids=lambda param: f"dtype={param}")
+def test_psnr(res, target_noise_variance, dtype):
+    vol = np.ones((res,) * 3, dtype=dtype)
+    g = grid_3d(res, normalized=False)
+    mask = g["r"] > res // 2
+    vol[mask] = 0
+
+    sim = Simulation(
+        vols=Volume(vol),
+        n=16,
+        amplitudes=1,
+        offsets=0,
+        dtype=dtype,
+        noise_adder=WhiteNoiseAdder(var=target_noise_variance),
+    )
+
+    psnr = sim.estimate_psnr(units="dB")
+    logger.debug(f"PSNR target={target_noise_variance} L={sim.L} {sim.dtype} {psnr}")
+    assert np.isclose(psnr, -10 * np.log10(target_noise_variance), rtol=0.01)
+
+
+@pytest.mark.parametrize(
+    "target_noise_variance", VARS, ids=lambda param: f"var={param}"
+)
+def test_from_snr_white(sim_fixture, target_noise_variance):
+    """
+    Test that prescribing noise directly by var and  by `from_snr`,
+    are close for a variety of paramaters.
+    """
+
+    # First add an explicit amount of noise to the base simulation,
+    sim_fixture.noise_adder = WhiteNoiseAdder(var=target_noise_variance)
+    # and compute the resulting snr of the sim.
+    target_snr = sim_fixture.estimate_snr()
+
+    # Attempt to create a new simulation at this `target_snr`
+    # For unit testing, we will use `sim_fixture`'s volume,
+    #   but the new Simulation instance should yield different projections.
+    sim_from_snr = Simulation.from_snr(
+        target_snr,
+        vols=sim_fixture.vols,  # Force the previously generated volume.
+        n=sim_fixture.n,
+        offsets=0,
+        amplitudes=1.0,
+        # Excercise the `sample_n` and shuffle branches
+        angles=uniform_random_angles(sim_fixture.n, dtype=sim_fixture.dtype),
+        sample_n=sim_fixture.n - 1,
+    )
+
+    # Check we're within 5% of explicit target
+    logger.info(
+        "sim_from_snr.noise_adder.noise_var, target_noise_variance ="
+        f" {sim_from_snr.noise_adder.noise_var}, {target_noise_variance}"
+    )
+    assert np.isclose(
+        sim_from_snr.noise_adder.noise_var, target_noise_variance, rtol=0.05
+    )
+
+    # Compare with WhiteNoiseEstimator consuming sim_from_snr
+    noise_estimator = WhiteNoiseEstimator(sim_from_snr, batchSize=512)
+    est_noise_variance = noise_estimator.estimate()
+    logger.info(
+        "est_noise_variance, target_noise_variance ="
+        f" {est_noise_variance}, {target_noise_variance}"
+    )
+
+    # Check we're within 5%
+    assert np.isclose(est_noise_variance, target_noise_variance, rtol=0.05)
+
+
+@pytest.mark.parametrize(
+    "target_noise_variance", VARS, ids=lambda param: f"var={param}"
+)
+def test_from_snr_pink_iso(sim_fixture, target_noise_variance):
+    """
+    Test that prescribing noise directly by var and  by `from_snr`,
+    are close for a variety of paramaters.
+    """
+
+    # Create the custom noise function and associated Filter
+    def pinkish_spectrum(x, y):
```

Comment:
May want to move this out to a separate function.

Reply:
I didn't want to curry the `target_noise_variance`... pulled it out, but end up with `lambda` functions.

Reply:
Scratch that, I was able to multiple with a scalar filter. Not sure its nicer 😇 .

Reply:
I see. It's less redundant at least. :+1: